hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/pubsub/PubSubMessageSenderProvider.java

```diff
@@ -0,0 +1,143 @@
+package pl.allegro.tech.hermes.consumers.consumer.sender.pubsub;
+
+import com.google.api.gax.batching.BatchingSettings;
+import com.google.api.gax.core.ExecutorProvider;
+import com.google.api.gax.core.FixedExecutorProvider;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.threeten.bp.Duration;
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.common.config.ConfigFactory;
+import pl.allegro.tech.hermes.common.config.Configs;
+import pl.allegro.tech.hermes.consumers.consumer.sender.MessageSender;
+import pl.allegro.tech.hermes.consumers.consumer.sender.ProtocolMessageSenderProvider;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class PubSubMessageSenderProvider implements ProtocolMessageSenderProvider {
+
+    public static final String SUPPORTED_PROTOCOL = "pubsub";
+
+    private static final Logger logger = LoggerFactory.getLogger(PubSubMessageSenderProvider.class);
+
+    private final PubSubSenderTargetResolver resolver;
+    private final ConfigFactory configFactory;
+    private final PubSubCredentialsProvider credentialsProvider;
+    private final PubSubMessages messageCreator;
+    private final ConcurrentMap<Subscription, Publisher> publishersCache = new ConcurrentHashMap<>();
```

Comment:
I think we have two options here:
1. We will fix the current behaviour regarding the `MessageSender#stop` method so that it is called on a subscription assignment removal. It should be easy, probably we should just call the `stop` in `ConsumerMessageSender::shutdown`.
2. We will cache `Publishers` per `PubSubSenderTarget` in a bounded data structure. I think this would be the most optimal approach, but it probably requires more work. (I assumed here that we can use the same `Publisher` for different Hermes subscriptions with the same endpoint.)

Reply:
Indeed. Apart from that: does Hermes create many `MessageSender` instances for one subscription (on a single node)? I assumed this statement is true, so I didn't want to create many `Publisher`'s. 

But you brought some problem to my attention. Unfortunately we have to create one publisher for one target topic, so the cache should have `GooglePubSubSenderTarget` as a key. You are able to post messages from many kafka topics to a single PubSub topic via one instance of `Publisher` then.

Reply:
I've created an alternative implementation of the cache/pool: https://github.com/allegro/hermes/commit/dd52031d79eb7cf2dc5178cbb9736170ed2fb59c. It can't grow indefinitely because it may have as many entries as the number of subscriptions assigned to the consumer node (which is limited). Additionally, unlike the Guave cache, it doesn't require tuning (setting parameters such as  `maximumSize` and `expireAfterAccess`). So it should be easier to use.

What do you think?

Reply:
I like it. It's definitely easier to maintain. I was not aware of the existence of the limit of subscriptions assigned to the consumer.